components/gantry/oneaxis/setup.go

```diff
@@ -0,0 +1,390 @@
+	}
+
+	switch len(conf.LimitSwitchPins) {
+	case 0:
+		// homes the motor and finds limits using the motor's position reporting
+		if err = oAx.homeEncoder(ctx); err != nil {
+			return nil, err
+		}
+		return oAx, nil
+	default:
+		// sets up a gantry that requires polling of external sensors, in this case limit switches
+		// to find position limits and gantry range
+		g, err := newLimitSwitchGantry(ctx, cfg, logger, oAx)
+		if err != nil {
+			return nil, err
+		}
+
+		lAx, ok := g.(*limitSwitchGantry)
+		if !ok {
+			return nil, rdkutils.NewUnexpectedTypeError(&limitSwitchGantry{}, lAx)
+		}
+		if err = lAx.homeWithLimSwitch(ctx, conf.LimitSwitchPins); err != nil {
+			return nil, err
+		}
+		return g, nil
+	}
+}
+
+type limitSwitchGantry struct {
+	ctx context.Context
+	oAx *oneAxis
+
+	limitSwitchPins []string
+	limitHigh       bool
+	limitHitMu      sync.Mutex
+	limitsPolled    bool
+
+	mu     sync.Mutex
+	logger golog.Logger
+	generic.Unimplemented
+}
+
+// newOneAxis creates a new one axis gantry.
+func newLimitSwitchGantry(
+	ctx context.Context,
+	cfg config.Component,
+	logger golog.Logger,
+	oneAxis *oneAxis,
+) (gantry.LocalGantry, error) {
+	conf, ok := cfg.ConvertedAttributes.(*AttrConfig)
+	if !ok {
+		return nil, rdkutils.NewUnexpectedTypeError(conf, cfg.ConvertedAttributes)
+	}
+
+	if len(conf.LimitSwitchPins) > 2 {
+		return nil, errBadNumLimitSwitches(cfg.Name, len(conf.LimitSwitchPins))
+	}
+
+	g := &limitSwitchGantry{
+		ctx:             ctx,
+		limitSwitchPins: conf.LimitSwitchPins,
+		limitHigh:       *conf.LimitPinEnabled,
+		oAx:             oneAxis,
+		logger:          logger,
+		limitsPolled:    false,
+	}
+
+	if len(g.limitSwitchPins) == 1 && (g.oAx.mmPerRevolution == 0 || g.oAx.lengthMm == 0) {
+		return nil, errDimensionsNotFound(g.oAx.name, g.oAx.mmPerRevolution, g.oAx.mmPerRevolution)
+	}
+
+	g.limitHitChecker()
+
+	return g, nil
+}
+
+func (g *limitSwitchGantry) limitHitChecker() {
+	// lock the mutex associated with this monitoring function
+	g.limitHitMu.Lock()
+	// check if the state is false
+	if !g.limitsPolled {
+		return
+	}
+	// we know we are monitoring the limit switches in this gantry,
+	// return if the limitHits are already being checked
+	g.limitsPolled = true
+	g.limitHitMu.Unlock()
+
+	var errs error
+	for {
+		select {
+		case <-g.ctx.Done():
+			return
+		default:
+		}
+		for idx := range g.limitSwitchPins {
+			hit, err := g.limitHit(g.ctx, idx)
+			if hit || err != nil {
+				g.limitHitMu.Lock()
+				g.limitsPolled = false
+				g.limitHitMu.Unlock()
+				errs := multierr.Combine(errs, err)
+				// motor operation manager cancels running
+				err = g.oAx.motor.Stop(g.ctx, nil)
+				errs = multierr.Combine(errs, err)
+				if errs != nil {
+					g.logger.Error(errs)
+					return
+				}
+			}
+		}
+	}
+}
+
+func (g *limitSwitchGantry) limitHit(ctx context.Context, offset int) (bool, error) {
+	if offset < 0 || offset > 1 {
+		return true, errBadNumLimitSwitches(g.oAx.name, offset)
```

Comment:
If `offset` is 2, we're going to get a confusing error message about how it can be 0, 1, or 2 but not 2. I suspect you're returning the wrong error here, though I'm unsure what the right one is. or maybe the if statement should check `offset > 2` instead?